### PR TITLE
Warn against deployment/update/deletion of stack/lambda based on CodePipeline presence

### PR DESCRIPTION
**Pull request by kyoge**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

When deploying a stack / updating a lambda function / deleting a stack or lambda function, we should check for the presence of a CI/CD pipeline. If one is found we should inform the user to use their CI/CD infrastructure instead.

This only acts as a warning guard rail, they can then click ignore and continue using the toolkit to do the work.

## Related Issue(s)

#867 

## Testing

Tested manually:
- **Warning is displayed** when deleting a CloudFormation stack or Lambda function which is part of an AWS CodePipeline.
- **Warning is displayed** when updating code/configuration of a Lambda function which is part of an AWS CodePipeline.
- **Warning is displayed** when updating a CloudFormation stack (while deploying serverless application) which is part of an AWS CodePipeline.

- _**Warning is not displayed**_ when deleting a CloudFormation stack or Lambda function which is not part of an AWS CodePipeline.
- _**Warning is not displayed**_ when updating code/configuration of a Lambda function which is not part of an AWS CodePipeline.
- _**Warning is not displayed**_ when updating a CloudFormation stack (while deploying serverless application) which is not part of an AWS CodePipeline.
- _**Warning is not displayed**_ when creating a new Lambda function or CloudFormation stack.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt [hunk 1]
```diff
@@ -75,14 +78,26 @@ class DeployServerlessApplicationAction : AnActionWrapper(
...
+            continueDeployment(project, stackName, templateFile, stackDialog)
+        } else {
+            warnResourceOperationAgainstCodePipeline(project, stackName, stackId, ResourceType.CLOUDFORMATION_STACK, Operation.DEPLOY) {
+                continueDeployment(project, stackName, templateFile, stackDialog)
+            }
+        }
+    }
+
+    private fun continueDeployment(project: Project, stackName: String, templateFile: VirtualFile, stackDialog: DeployServerlessApplicationDialog) {
         val deployDialog = SamDeployDialog(
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: Indent size 4, continuation indent 8 are the kotlin standards

Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: This is following the same (indent size 4, continuation indent 8) kotlin standards. What is the ask here?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt [hunk 2]
```diff
@@ -121,6 +128,18 @@ class DeployServerlessApplicationDialog(
             view.stacks.selected()
         } ?: throw RuntimeException(message("serverless.application.deploy.validation.stack.missing"))
 
+    val stackId: String?
+        get() = if (view.createStack.isSelected) {
+            null
+        } else {
+            val stackName = view.stacks.selected()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: stackName?.let {
	stackNameToStackId[it]
}

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 3]
```diff
@@ -138,7 +139,7 @@ class ResourceSelector<T> : ComboBox<T>() {
         }
     }
 
-    fun values(): Set<T> = (0 until itemCount).map { getItemAt(it) }.toSet()
+    fun values(): List<T> = (0 until itemCount).map { getItemAt(it) }.toList()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Why is this changed?

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt [hunk 4]
```diff
@@ -48,31 +48,31 @@ class CloudFormationStackNodeTest {
...
         whenever(mockCfnClient.describeStackResources(any<DescribeStackResourcesRequest>())).thenReturn(
-            DescribeStackResourcesResponse.builder()
-                .stackResources(
-                    StackResource.builder().resourceType(LAMBDA_FUNCTION_TYPE).resourceStatus(ResourceStatus.CREATE_COMPLETE).logicalResourceId("processor").build(),
-                    StackResource.builder().resourceType(LAMBDA_FUNCTION_TYPE).resourceStatus(ResourceStatus.CREATE_COMPLETE).logicalResourceId("processor2").build(),
-                    StackResource.builder().resourceType("a dynamodb table").resourceStatus(ResourceStatus.CREATE_COMPLETE).build(),
-                    StackResource.builder().resourceType("an IAM role").resourceStatus(ResourceStatus.CREATE_COMPLETE).build()
-                )
-                .build()
+                DescribeStackResourcesResponse.builder()
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt]: Why are you changing this formatting?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt [hunk 5]
```diff
@@ -43,6 +43,7 @@ class DeployServerlessApplicationDialog(
     private val s3Client: S3Client = project.awsClient()
     private val cloudFormationClient: CloudFormationClient = project.awsClient()
     private val templateParameters = CloudFormationTemplate.parse(project, templateFile).parameters().toList()
+    private lateinit var stackNameToStackId: Map<String, String>
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Why not just store this mapping as a data class in the selector itself?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt (hunk 1 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: It looks to be 8, you have changed the formatting in several areas of the code base and I don't see anything that causes it to be wrong in the first place

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt (hunk 4 above)
Comment (kyoge) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt]: I've not changed the formatting, I just touched this file and the IDE reformatted it itself. I am using the default Kotlin code style settings (indent size 4, continuation indent 8) and I've never changed them.

Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt]: There are 2 Kotlin code styles in Intellij, make sure you aren't using the obsolete one.

Comment (kyoge) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt]: May be, they changed it in a recent update in IntelliJ IDEA. The parenthesis '(' is followed by a continuation indent.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt (hunk 1 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: Actually it is supposed to be 4 according to yole

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: https://github.com/JetBrains/kotlin/blob/189260c0134741a7e1fe2783193b34c8a241901a/.idea/codeStyles/Project.xml#L24

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+fun getCodePipelineArnForResource(project: Project, resourceArn: String, resourceTypeFilter: String): String? {
+    val client: ResourceGroupsTaggingApiClient = project.awsClient()
+
+    val tagFilter = TagFilter.builder().key(CODEPIPELINE_SYSTEM_TAG_KEY).build()
+    val request = GetResourcesRequest.builder().tagFilters(tagFilter).resourceTypeFilters(resourceTypeFilter).build()
+    val response = client.getResources(request)
+
+    for (resourceTagMapping in response.resourceTagMappingList().filterNotNull()) {
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Good catch! will fix.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import com.intellij.openapi.ui.showYesNoDialog
+import software.amazon.awssdk.services.resourcegroupstaggingapi.ResourceGroupsTaggingApiClient
+import software.amazon.awssdk.services.resourcegroupstaggingapi.model.GetResourcesRequest
+import software.amazon.awssdk.services.resourcegroupstaggingapi.model.TagFilter
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.resources.message
+
+const val CODEPIPELINE_SYSTEM_TAG_KEY = "aws:codepipeline:pipelineArn"
+
+enum class ResourceType(val value: String, val filter: String) {
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Actually, it's not a tagFilter, it's a resourceTypeFilter. If I rename it to resourceTypeFilter, it'd be kind of a duplication as the enum name is ResourceType already. Any other suggestion?

[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: tagResourceType then? It identifies the string that tagging service uses to identify a service.

We need comments and/or a better name

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: This is a leftover change suggested by @kiiadi in a previous PR. Returning list from here will support a caller who is interested in the order of items in future.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt (hunk 2 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Nice! I thought that I could use something concise like you suggested, it just didn't click to my mind at that time.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Okay... I would have just changed the comnbo box model to CollectionComboBoxModel and used `CollectionComboBoxModel#getList()`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt (hunk 5 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: I think selector shouldn't store this because this mapping is specific to stacks and the selector is a general purpose combo box storing not only stacks but also s3Buckets and iamRoles etc.

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: The selector just stores data, you can store anything in a combo box and just stick a custom renderer on it.

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: This is how we do IAM roles as well. Its a ResourceSelector<IamRole>

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt (hunk 4 above)
Comment (kyoge) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationStackNodeTest.kt]: So, I changed Kotlin code style in my IDE: Editor > Code Style > Kotlin > Set from... > Predefined Style > Kotlin style guide. Now, it seems to match the style of our codebase...hopefully.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt (hunk 1 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: I changed Kotlin code style in my IDE: Editor > Code Style > Kotlin > Set from... > Predefined Style > Kotlin style guide. It uses the indent size 4 after parenthesis now.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Actually tagFilter seems better to me now in the context of ResourceType. Renaming to tagFilter and adding comments as well.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Fixed.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt (hunk 5 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Done, using ResourceSelector<Stack> now like ResourceSelector<IamRole>.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: ^^ what was the resolution here? the main use of the `values` method is to perform an 'exists' function right? that's why `Set` is a better choice of data-structure - why should the caller ever need to know about the order of this? That's really a view (and thus the `model`'s) concern.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 8]
```diff
@@ -81,15 +83,16 @@ class ResourceSelector<T> : ComboBox<T>() {
                 null
             } ?: return@executeOnPooledThread
 
-            ApplicationManager.getApplication().invokeLater({
+            runInEdt(ModalityState.any()) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: maybe we should have a ktlint rule here if we want to be consistent around how we run on the UI thread (e.g. `ApplicationManager` vs simple `runInEdt`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        }
+    }
+}
+
+fun getCodePipelineArnForResource(project: Project, resourceArn: String, resourceTypeFilter: String): String? {
+    val client: ResourceGroupsTaggingApiClient = project.awsClient()
+
+    val tagFilter = TagFilter.builder().key(CODEPIPELINE_SYSTEM_TAG_KEY).build()
+
+    var paginationToken: String? = ""
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: why are we handling pagination manually? The SDK exposes paginated APIs for us : https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/resourcegroupstaggingapi/ResourceGroupsTaggingApiClient.html#getResourcesPaginator-java.util.function.Consumer-

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/deploy/DeploySamApplicationValidatorTest.kt [hunk 10]
```diff
@@ -31,8 +31,8 @@ class DeploySamApplicationValidatorTest {
     fun wireMocksTogetherWithValidOptions() {
 
         val parameters = listOf<Parameter>(
-                TestParameter("param1", "value1"),
-                TestParameter("param2", "value2")
+            TestParameter("param1", "value1"),
```
Comment (kiiadi) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/deploy/DeploySamApplicationValidatorTest.kt]: spacing

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import software.amazon.awssdk.services.resourcegroupstaggingapi.model.TagFilter
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.resources.message
+
+const val CODEPIPELINE_SYSTEM_TAG_KEY = "aws:codepipeline:pipelineArn"
+
+/**
+ * @property tagFilter used by ResourceGroupsTaggingApi to filter response by ResourceType (resourceTypeFilter)
+ */
+enum class ResourceType(val value: String, val tagFilter: String) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: This seems like a very generic name for something so specialized.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: The current caller doesn't need to know the order of elements, so Set is better. I used List initially but changed it to Set upon your suggestion in #899 where later you suggested that `the values() method should still return a list - other resource selectors may care about ordering (and may contain duplicates) which you will lose if you move this to a Set.`

Well, in the latest revision, I moved from using `ResourceSelector<String>` to `ResourceSelector<Stack>` where Stack is a data class containing stack name and id. Now I use `values()` method not to perform an 'exists' function but to find an element matching a condition, so Set doesn't give me any advantage.

I think I'll just move to CollectionComboBoxModel as Austin suggested and get rid of this method.

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: you should be able to use `software.aws.toolkits.jetbrains.utils.ui.UiUtils#find` that allows you to find an item within a combobox model ...

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 12]
```diff
@@ -54,18 +56,37 @@ class CloudFormationServiceNode(project: Project) : AwsExplorerServiceRootNode(p
     }
 }
 
-class CloudFormationStackNode(project: Project, val stackName: String, private val stackStatus: StackStatus, val stackId: String) :
-    AwsExplorerResourceNode<String>(project, CloudFormationClient.SERVICE_NAME, stackName, AwsIcons.Resources.CLOUDFORMATION_STACK),
+class CloudFormationStackNode(
+    project: Project,
+    val stackName: String,
+    private val stackStatus: StackStatus,
+    val stackId: String
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: So, this is formatted by the official Kotlin formatter. Austin also commented about formatting which I replied above, pasting my reply again here:

I see there are only 2 predefined Kotlin code styles in preferences:

1. **Kotlin obsolete IntelliJ IDEA codestyle**: This is the default style I had in my IDE which uses continuation indent (sized 8) after parenthesis '(' which doesn't look very nice.

2. **Kotlin style guide**: Following https://kotlinlang.org/docs/reference/code-style-migration-guide.html, I changed code style in my IDE (Editor > Code Style > Kotlin > Set from... > Predefined Style > Kotlin style guide) and added 'kotlin.code.style=official' property to the gradle.properties file so that all of us use the same official code style which makes formatting consistent with the Kotlin Coding Conventions. This new style uses the indent size 4 after parenthesis. The continuation indent size is also 4 in this style, so there will be some formatting changes in the files I am touching.

**Note**: if we prefer using a different code style than these predefined ones, `.idea/codeStyle` folder should be committed so that all project developers could share the same code style.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt (hunk 9 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Oh, nice! I didn't know that such paginated APIs exist.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt (hunk 11 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: I'll rename it to TaggingResourceType.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt [hunk 13]
```diff
@@ -71,28 +71,30 @@ class DeployServerlessApplicationDialog(
...
+                .sortedWith(String.CASE_INSENSITIVE_ORDER)
+                .toList()
         }
 
         view.createS3BucketButton.addActionListener {
             val bucketDialog = CreateS3BucketDialog(
-                    project = project,
-                    s3Client = s3Client,
-                    parent = view.content
+                project = project,
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: What is the issue about spacing here? Austin commented that indent size should be 4 after parenthesis, that's why I made this change in formatting to use indent (size 4) in place of continuation indent (size 8). Do we follow a common code style which could be used by all of us in our IDEs to be consistent?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt [hunk 14]
```diff
@@ -118,12 +120,24 @@ class DeployServerlessApplicationDialog(
...
+                // DeploySettings doesn't store stack id because it doesn't have access to stack id
+                // at times when deployment happens with createStack selected
+                stack.id ?: view.stacks.values().find { it.name == stack.name }?.id
+            }
+        }
+
     val bucket: String
         get() = view.s3Bucket.selected()
-                ?: throw RuntimeException(message("serverless.application.deploy.validation.s3.bucket.empty"))
+            ?: throw RuntimeException(message("serverless.application.deploy.validation.s3.bucket.empty"))
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Same as above

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 8 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: I tried adding a BannedPattern in the existing BannedPatternRule but it didn't work. ktlint doesn't seem to even catch the existing banned patterns defined in this rule. I tried using the existing banned patterns in my code and reran ktlint, it succeeds.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeleteFunctionAction.kt [hunk 15]
```diff
@@ -5,10 +5,13 @@ package software.aws.toolkits.jetbrains.services.lambda.actions
...
 import software.aws.toolkits.jetbrains.services.lambda.LambdaFunctionNode
+import software.aws.toolkits.jetbrains.utils.ResourceType
 import software.aws.toolkits.resources.message
 
 class DeleteFunctionAction : DeleteResourceAction<LambdaFunctionNode>(message("lambda.function.delete.action")) {
     override fun performDelete(selected: LambdaFunctionNode) {
         selected.client.deleteFunction { it.functionName(selected.functionName()) }
     }
+
+    override fun getResourceType() = ResourceType.LAMBDA_FUNCTION
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeleteFunctionAction.kt]: Pushed it to abstract class as a property.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 16]
```diff
@@ -137,10 +159,13 @@ class CloudFormationStackNode(project: Project, val stackName: String, private v
...
-class DeleteCloudFormationStackAction : DeleteResourceAction<CloudFormationStackNode>(message("cloudformation.stack.delete.action")) {
+class DeleteCloudFormationStackAction :
+    DeleteResourceAction<CloudFormationStackNode>(message("cloudformation.stack.delete.action")) {
     override fun performDelete(selected: CloudFormationStackNode) {
         val client: CloudFormationClient = AwsClientManager.getInstance(selected.nodeProject).getClient()
         client.deleteStack { it.stackName(selected.stackName) }
         client.waitForStackDeletionComplete(selected.stackName)
     }
+
+    override fun getResourceType() = ResourceType.CLOUDFORMATION_STACK
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: Renamed it to taggingResourceType to be more specific and pushed it to the abstract class as a property.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 17]
```diff
@@ -54,18 +56,37 @@ class CloudFormationServiceNode(project: Project) : AwsExplorerServiceRootNode(p
...
+    AwsExplorerResourceNode<String>(
+        project,
+        CloudFormationClient.SERVICE_NAME,
+        stackName,
+        AwsIcons.Resources.CLOUDFORMATION_STACK
+    ),
     AwsNodeAlwaysExpandable {
     override fun resourceType() = "stack"
 
+    override fun resourceArn() = stackId
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: `stackId` has to be there as a class property in `CloudFormationStackNode ` because it is initialized at the time of object construction:
```
val nodes = response.stacks().filterNotNull().asSequence()
            .filter { it.stackStatus() !in DELETING_STACK_STATES }
            .sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.stackName() })
            .map { CloudFormationStackNode(nodeProject, it.stackName(), it.stackStatus(), it.stackId()) }
            .toList()
```

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 18]
```diff
@@ -54,18 +56,37 @@ class CloudFormationServiceNode(project: Project) : AwsExplorerServiceRootNode(p
...
+
+    override fun displayName() = stackName
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: Same as above

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Thanks, I could use UiUtils#find. Removed the `values()` method from ResourceSelector as I don't need it now.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 19]
```diff
@@ -54,18 +56,37 @@ class CloudFormationServiceNode(project: Project) : AwsExplorerServiceRootNode(p
...
     private val cfnClient: CloudFormationClient = project.awsClient()
 
     /**
      * CloudFormation Stack Nodes do not immediately query for stack resources.
      * We wait until node will be expanded before querying, reducing risk of triggering TPS limits.
      */
-    private val noResourcesChildren: Collection<AbstractTreeNode<Any>> = listOf(AwsExplorerEmptyNode(project, message("explorer.stack.no.serverless.resources"))).filterIsInstance<AbstractTreeNode<Any>>()
+    private val noResourcesChildren: Collection<AbstractTreeNode<Any>> = listOf(
+        AwsExplorerEmptyNode(
+            project,
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt [hunk 20]
```diff
@@ -137,10 +159,13 @@ class CloudFormationStackNode(project: Project, val stackName: String, private v
     }
 }
 
-class DeleteCloudFormationStackAction : DeleteResourceAction<CloudFormationStackNode>(message("cloudformation.stack.delete.action")) {
+class DeleteCloudFormationStackAction :
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationExplorerNodes.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt [hunk 21]
```diff
@@ -157,5 +187,6 @@ class DeployServerlessApplicationAction : AnActionWrapper(
         }
     }
 
-    private fun validateTemplateFile(project: Project, templateFile: VirtualFile): String? = project.validateSamTemplateLambdaRuntimes(templateFile)
+    private fun validateTemplateFile(project: Project, templateFile: VirtualFile): String? =
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/UpdateFunctionAction.kt [hunk 22]
```diff
@@ -31,16 +34,24 @@ abstract class UpdateFunctionAction(private val mode: EditFunctionMode, title: S
...
+            ) {
                 EditFunctionDialog(project, lambdaFunction, mode = mode).show()
             }
         }
     }
 }
 
-class UpdateFunctionConfigurationAction : UpdateFunctionAction(EditFunctionMode.UPDATE_CONFIGURATION, message("lambda.function.updateConfiguration.action"))
+class UpdateFunctionConfigurationAction :
+    UpdateFunctionAction(EditFunctionMode.UPDATE_CONFIGURATION, message("lambda.function.updateConfiguration.action"))
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/UpdateFunctionAction.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt [hunk 23]
```diff
@@ -195,9 +209,9 @@ class DeploySamApplicationValidator(private val view: DeployServerlessApplicatio
         val parameters = view.templateParameters
 
         val unsetParameters = parameters.entries
-                .filter { it.value.isNullOrBlank() }
-                .map { it.key }
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt]: Reverted formatting changes.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 24]
```diff
@@ -120,11 +123,15 @@ class ResourceSelector<T> : ComboBox<T>() {
...
                 }
-            }, ModalityState.any())
+            }
         }
     }
 
-    fun toValidationInfo(loading: String = message("loading_resource.loading"), failed: String = message("loading_resource.failed"), notSelected: String): ValidationInfo? {
+    fun toValidationInfo(
+        loading: String = message("loading_resource.loading"),
+        failed: String = message("loading_resource.failed"),
```
[the remark this answers is not in the document]
Comment (kyoge) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Reverted formatting changes.

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/deploy/DeploySamApplicationValidatorTest.kt (hunk 10 above)
Comment (kyoge) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/deploy/DeploySamApplicationValidatorTest.kt]: Reverted formatting changes.

**Comment by abrooksv** [comment]
Looks to have ktlint violations

**Comment by kyoge** [comment]
> Looks to have ktlint violations

Looking at it.
